cmake/godotcpp.cmake

```diff
@@ -36,205 +115,182 @@ function( godotcpp_options )
-
-    # Create the correct name (godot.os.build_type.system_bits)
-    string(TOLOWER "${CMAKE_SYSTEM_NAME}" SYSTEM_NAME)
-    string(TOLOWER "${CMAKE_BUILD_TYPE}" BUILD_TYPE)
-
-    if(ANDROID)
-        # Added the android abi after system name
-        set(SYSTEM_NAME ${SYSTEM_NAME}.${ANDROID_ABI})
-
-        # Android does not have the bits at the end if you look at the main godot repo build
-        set(OUTPUT_NAME "godot-cpp.${SYSTEM_NAME}.${BUILD_TYPE}")
+    ### Use the arch from the toolchain if it isn't set manually
+    if( GODOT_ARCH )
+        set(SYSTEM_ARCH ${GODOT_ARCH})
     else()
-        set(OUTPUT_NAME "godot-cpp.${SYSTEM_NAME}.${BUILD_TYPE}.${BITS}")
+        godot_arch_map( SYSTEM_ARCH ${CMAKE_SYSTEM_PROCESSOR} )
     endif()
 
-    set_target_properties(${PROJECT_NAME}
-            PROPERTIES
-            CXX_EXTENSIONS OFF
-            POSITION_INDEPENDENT_CODE ON
-            CXX_VISIBILITY_PRESET ${GODOT_SYMBOL_VISIBILITY}
-            ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/bin"
-            LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/bin"
-            RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/bin"
-            OUTPUT_NAME "${OUTPUT_NAME}"
-    )
+    ### Define our godot-cpp library targets
+    foreach ( TARGET_NAME template_debug template_release editor )
+
+        # Useful genex snippits used in subsequent genex's
+        set( IS_RELEASE "$<STREQUAL:${TARGET_NAME},template_release>")
+        set( IS_DEV "$<BOOL:${GODOT_DEV_BUILD}>")
+        set( DEBUG_FEATURES "$<OR:$<STREQUAL:${TARGET_NAME},template_debug>,$<STREQUAL:${TARGET_NAME},editor>>" )
+        set( HOT_RELOAD "$<IF:${HOT_RELOAD-UNSET},$<NOT:${IS_RELEASE}>,$<BOOL:${GODOT_USE_HOT_RELOAD}>>" )
+
+        # the godot-cpp.* library targets
+        add_library( ${TARGET_NAME} STATIC ${EXCLUDE} )
+        add_library( godot-cpp::${TARGET_NAME} ALIAS ${TARGET_NAME} )
+
+        file( GLOB_RECURSE GODOTCPP_SOURCES LIST_DIRECTORIES NO CONFIGURE_DEPENDS src/*.cpp )
+
+        target_sources( ${TARGET_NAME}
+                PRIVATE
+                ${GODOTCPP_SOURCES}
+                ${GENERATED_FILES_LIST}
+        )
+
+        target_include_directories( ${TARGET_NAME} ${GODOT_SYSTEM_HEADERS_ATTRIBUTE} PUBLIC
+                include
+                ${CMAKE_CURRENT_BINARY_DIR}/gen/include
+                ${GODOT_GDEXTENSION_DIR}
+        )
+
+        set_target_properties( ${TARGET_NAME}
+                PROPERTIES
+                CXX_STANDARD 17
+                CXX_EXTENSIONS OFF
+                CXX_VISIBILITY_PRESET ${GODOT_SYMBOL_VISIBILITY}
+
+                COMPILE_WARNING_AS_ERROR ${GODOT_WARNING_AS_ERROR}
+                POSITION_INDEPENDENT_CODE ON
+                BUILD_RPATH_USE_ORIGIN ON
+
+                PREFIX lib
+                OUTPUT_NAME "${PROJECT_NAME}.${SYSTEM_NAME}.${TARGET_NAME}.${SYSTEM_ARCH}"
```

Comment:
It's line 272 that causes the offence.
>        ARCHIVE_OUTPUT_DIRECTORY "$<1:${CMAKE_SOURCE_DIR}/bin>"

I agree, the build artifact location shouldn't be in the source tree. And final artifact location should be specified by an install directive. But, I am trying to tightly match the output of the scons build system, perhaps to a fault, where in tools/godotcpp.py:562

>        library = env.StaticLibrary(target=env.File("bin/%s" % library_name), source=sources)

SCons places the build artifacts in that location. Granted SCons isn't setup afaik to do out of source builds so it isnt a problem there. 

So we have a situation where perfect might be the enemy of the good. I'd like to get this merged. There is your PR #1418 in which we could adress this?

Reply:
So long as this isn't a regression with regard to what the previous cmake configuration allowed, I'd also prefer to merge what we have and continue to improve in further PRs.

Is it a regression? I just tried building via cmake on `master`, and it puts the static library into `cmake-build/bin/` rather than the top-level `bin/`, so it does seem like a regression.

Reply:
As for matching with SCons, using CMAKE_BINARY_DIR and running cmake in the top level directory (rather than in a separate build directory) will still place the build outputs in the same place as the SCons build.

Reply:
It is more correct to use binary directory. If you think about the scons build as having these concepts then it's binary directory is the same as the source directory, so in that way it isn't a regression like @tynuf said. 

Projects that rely on cmake dont care where the output is because they are told by cmake itself. projects that use scons will find the build artifacts in the directory they always have. and if you're mixing build systems, then you're used to dealing with the differences.

I've made the change as requested by @ytnuf as I expect it will be forced by upcoming changes.